app/controllers/committee/agenda_types_controller.rb

```diff
@@ -0,0 +1,50 @@
+# frozen_string_literal: true
+
+module Committee
+  class AgendaTypesController < ApplicationController
+    include Pagy::Backend
+
+    before_action :set_agenda_type, only: %i[edit update destroy]
+
+    def index
+      @pagy, @agenda_types = if params[:term].present?
```

Comment:
Bu kalıp yani aşağıdaki form sadece bu PR'da 3 yerde geçiyor.

```
@pagy, .* = if params[:term].present?
              pagy(.*.search(params[:term])
            else
              pagy(.*)
            end 
```

Sadece burada geçseydi göz ardı ederdim ama kodda aşağıdaki aramayı yaptım:

```sh
grep -ERA 3 '^\s*@pagy, .* = if params\[:term\]' .
```

Sonuç: halihazırda bu kalıp 7 yerde geçiyor.  Bu PR'ı da eklersek 10 ediyor, ki bence bu rakam burada durmaz, gelecekte daha da büyür.  (İlk match (smart_search, dynamic_search) kalıba tam uymuyor, ama küçük bir çabayla onu da uydururuz).

```
./app/controllers/units_controller.rb:    @pagy, @units = if params[:term].present?
./app/controllers/units_controller.rb-                      pagy(smart_search(units))
./app/controllers/units_controller.rb-                    else
./app/controllers/units_controller.rb-                      pagy(dynamic_search(units))
--
./app/controllers/users_controller.rb:    @pagy, @users = if params[:term].present?
./app/controllers/users_controller.rb-                      pagy(User.all.search(params[:term]))
./app/controllers/users_controller.rb-                    else
./app/controllers/users_controller.rb-                      pagy(User.all)
--
./app/controllers/calendar/calendar_titles_controller.rb:      @pagy, @calendar_titles = if params[:term].present?
./app/controllers/calendar/calendar_titles_controller.rb-                                  pagy(CalendarTitle.search(params[:term]))
./app/controllers/calendar/calendar_titles_controller.rb-                                else
./app/controllers/calendar/calendar_titles_controller.rb-                                  pagy(CalendarTitle.all)
--
./app/controllers/locations/countries_controller.rb:      @pagy, @countries = if params[:term].present?
./app/controllers/locations/countries_controller.rb-                            pagy(countries.search(params[:term]))
./app/controllers/locations/countries_controller.rb-                          else
./app/controllers/locations/countries_controller.rb-                            pagy(countries)
--
./app/controllers/locations/countries_controller.rb:      @pagy, @cities = if params[:term].present?
./app/controllers/locations/countries_controller.rb-                         pagy(@country.cities.search(params[:term]))
./app/controllers/locations/countries_controller.rb-                       else
./app/controllers/locations/countries_controller.rb-                         pagy(@country.cities)
--
./app/controllers/locations/cities_controller.rb:      @pagy, @districts = if params[:term].present?
./app/controllers/locations/cities_controller.rb-                            pagy(districts.search(params[:term]))
./app/controllers/locations/cities_controller.rb-                          else
./app/controllers/locations/cities_controller.rb-                            pagy(districts)
--
./app/controllers/languages_controller.rb:    @pagy, @languages = if params[:term].present?
./app/controllers/languages_controller.rb-                          pagy(languages.search(params[:term]))
./app/controllers/languages_controller.rb-                        else
./app/controllers/languages_controller.rb-                          pagy(languages)
```
Küçük bir DRY şeysi yararlı olabilir.  Yazması benden, (isimlendirmeleri de gözden geçirerek) uygun yere monte etmesi sizden olsun:

```ruby
  def pagy_on(subject)
    @pagy, found = pagy(params[:term].present? ? subject.search(params[:term]) : subject)
    found
  end
```

Örneğin buradaki kullanımı:

```ruby
@agenda_types = pagy_on(AgendaType.all)
```

Böyle bir değişiklik yaparsanız 10x5 = 50 satır, 10x1 + 3 = 13 satıra iner (37 satır silinir).  Nereye konulabilir?  Bu bir helper değil aslında.  Zaman bulursam ben de bakarım.

P.S. Gereği isteğe bağlı bir comment olarak yazıyorum bunu, "change request" olarak kayda girmesin.

Reply:
@msdundar Sanıyorum bunun yeri: `app/controllers/concerns` dizini.  Orada zaten bir kaç hazır concern buldum :+1:   `app/controllers/concerns/pagy_backend_with_helpers.rb` gibi bir isimle gidersek aşağıdaki gibi bir şey uyar.  Test etmeden yazıyorum.  Concern mekaniğine hakim değilim, `reference_resource.rb`'deki mekanikten esinlenerek yazdım.

```ruby
# frozen_string_literal: true

module PagyBackendWithHelpers
  extend ActiveSupport::Concern
  include Pagy::Backend

  included do
    def pagy_by_search(collection)
      @pagy, items = pagy(
        (term = params[:term]).present? ? collection.search(term) : collection
      )
      items
    end
  end
end
```

Kullanımı:

```ruby
include PagyBackendWithHelpers # include Pagy::Backend yerine
...
@agenda_types = pagy_by_search(AgendaType.all)
```

İsimlendirme hakkında bir kaç düşünce:

- `pagy_by_search ismi search eylemini daha güzel yansıtıyor

- metoddaki `collection` ve `items` isimlerini Pagy lingo'dan aldım.  ([Kaynak](https://github.com/ddnexus/pagy/blob/master/lib/pagy/backend.rb))

- concern ismi doğrudan `PagyBackendWithSearch` gibi bir şey de olabilirdi, ama bence bu modüle yeni eklemeler yapılabilir, o nedenle ismi jenerik tuttum.  isimde helpers geçmesinden rahatsızım :-P alternatif isim `PagyBackendEnhanced` olabilir.

Reply:
P.S.  Bu concern'de bilinmesi gereken bir husus var.  `include Pagy::Backend` yapan bir başka concern de aynı dosyada kullanılırsa (ör. `ReferenceResource`) iki defa  include yapmış oluyoruz.  Şu an böyle bir durum yok, ama gelecekte olabilir.  Böyle bir durumda concern'lerden include'u çıkarıp çalışmak uygun olabilir (bunun ismi de `PagyBackendHelpers` gibi bir şeye dönüşür).

Reply:
Yol üzerinde Rails 5 ile gelen bir yenilik şu okumayı yaptım: https://blog.bigbinary.com/2016/06/26/rails-add-helpers-method-to-ease-usage-of-helper-modules-in-controllers.html

Siz de okuyun lütfen.

Reply:
Concern'ün son hali: https://gist.github.com/fbe323237a2bea59cd50a686190125b4